src/gluonts/dataset/hierarchical.py

```diff
@@ -0,0 +1,217 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+
+# Standard library imports
+from typing import Optional
+
+# Third-party imports
+import numpy as np
+import pandas as pd
+
+# First-party imports
+from gluonts.dataset.pandas import PandasDataset
+
+
+class HierarchicalTimeSeries:
+    r"""
+    Class for representing hierarchical time series.
+
+    The hierarchy is represented by the standard aggregation matrix `S`.
+    The time series at the bottom (leaf) level of the hierarchy
+    (`ts_at_bottom_level`) are assumed to be given by the columns of
+    a single pandas dataframe.
+
+    The ordering of columns of `ts_at_bottom_level` should be consistent
+    with the ordering of the columns of `S`.
+
+    Parameters
+    ----------
+    ts_at_bottom_level
+        A single pandas dataframe whose columns are the time series
+        corresponding to the leaves of the hierarchy.
+    S
+        Summation or aggregation matrix whose ordering should be consistent
+        with the orderig of the columns of `ts_at_all_levels`.
+        In particular, the bottom `k x k` sub-matrix should be identity matrix,
+        where `k` is the number of leaves of the hierarchy.
```

Comment:
If this is strictly required, then maybe it's fine to omit the bottom `k x k` submatrix and leave it implicit?

That would mean that `S` encodes the non-leaf nodes in the hierarchy only. I understand this may go against common mathematical notation, but it seems better on the coding side of things.

Reply:
Didn't want to introduce yet another notation and users typically have this matrix available for their own processing tasks as well: multiplying the forecasts at bottom level by this matrix gives them "all" forecasts in one go. There is a good reason for the introduction of `S` in the literature and from the user-perspective, it makes sense to use this standard matrix.

Reply:
I still think it's strange to require some input if part of it is expected to have specific values. But we can go on with this and maybe revisit it later.